### PR TITLE
SDXL: Add graph mode training for dreambooth

### PR DESCRIPTION
**Pull request by Fzilan**
1. add graph mode
2. add filtering weight
3. **modify lora layer: q,v -> q,v,k, out**
4. fix bug

### CODE REVIEW

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 1]
```diff
@@ -445,7 +452,6 @@ def init_sampling(
     num_cols=None,
     sampler="EulerEDMSampler",
     guider="VanillaCFG",
-    guidance_scale=5.0,
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/helpers.py:448]: 保留guidance_scale参数

File: examples/stable_diffusion_xl/gm/models/diffusion.py [hunk 2]
```diff
@@ -251,7 +250,7 @@ def do_sample(
         randn = Tensor(np.random.randn(*shape), ms.float32)
 
         print("Sample latent Starting...")
-        samples_z = sampler(self, randn, cond=c, uc=uc, adapter_states=adapter_states)
+        samples_z = sampler(self, randn, cond=c, uc=uc)
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/models/diffusion.py:253]: 保留

File: examples/stable_diffusion_xl/gm/models/diffusion.py [hunk 3]
```diff
@@ -352,10 +351,10 @@ def do_img2img(
 
 class DiffusionEngineDreamBooth(DiffusionEngine):
     def __init__(self, prior_loss_weight=1.0, *args, **kwargs):
-        super().__init__(self, *args, **kwargs)
+        super().__init__(*args, **kwargs)
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/models/diffusion.py:354]: 这个地方为什么要删掉self？

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 4]
```diff
@@ -125,14 +120,7 @@ def set_default(args):
...
 
-def create_model(
-    config: DictConfig,
-    checkpoints: Union[str, List[str]] = "",
-    freeze: bool = False,
-    load_filter: bool = False,
-    param_fp16: bool = False,
-    amp_level: Literal["O0", "O1", "O2", "O3"] = "O0",
-):
+def create_model(config, checkpoints=None, freeze=False, load_filter=False, param_fp16=False, amp_level="O0"):
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/helpers.py:123]: 保留原来写法

File: examples/stable_diffusion_xl/gm/models/diffusion.py (hunk 3 above)
Comment (hadipash) [reply examples/stable_diffusion_xl/gm/models/diffusion.py:354]: It's correct. It's either `DiffusionEngine.__init__(self, *args, **kwargs)` (which is not good), or `super().__init__(*args, **kwargs)` (preferred way).

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py [hunk 5]
```diff
@@ -29,6 +29,52 @@ def construct(self, x, noised_input, sigmas, w, concat, context, y):
         return self.scaler.scale(loss)
 
 
+class LatentDiffusionWithLossDreamBooth(nn.Cell):
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/models/trainer_factory.py:32]: Use inheritance and just call `super(...)` instead of `self._shared_step(...)`

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py [hunk 6]
```diff
@@ -98,3 +144,73 @@ def construct(self, x, *tokens):
         loss, _, overflow = self.ldm_with_loss_grad(x, noised_input, sigmas, w, concat, context, y)
 
         return loss, overflow
+
+
+class TrainOneStepCellDreamBooth(nn.Cell):
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/models/trainer_factory.py:149]: I feel like creating `TrainOneStepCell` for each model is suboptimal. Maybe you can refer to `examples.stable_diffusion_v2.ldm.modules.train.trainer.TrainOneStepWrapper` to see how we utilize `TrainOneStepCell` for all SD variations.

**Comment by SamitHuang** [comment]
Please update the readme file as well @Fzilan

File: examples/stable_diffusion_xl/configs/training/sd_xl_base_finetune_dreambooth_lora_910b.yaml [hunk 7]
```diff
@@ -133,19 +133,19 @@ optim:
 
 data:
     per_batch_size: 1
-    total_step: 12000
-    num_parallel_workers: 6
+    total_step: 8000
```
Comment (SamitHuang) [inline examples/stable_diffusion_xl/configs/training/sd_xl_base_finetune_dreambooth_lora_910b.yaml:136]: batch size=1 is too small for good optimization. Suggest to improve via gradient accumulation or stronger recompute or parallel training

File: examples/stable_diffusion_xl/gm/models/trainer_factory.py (hunk 6 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/models/trainer_factory.py:167]: marked as todo

File: examples/stable_diffusion_xl/dreambooth_finetune.md [hunk 8]
```diff
@@ -41,15 +41,23 @@ dir
...
 <p align="center">
-  <em> Figure 1. The five images from the subject dog for finetuning. </em>
+  <em> Figure 1. dog example: the five images from the subject dog for finetuning. </em>
 </p>
 
+<p align="center">
+  <img src="https://github.com/mindspore-lab/mindone/assets/33061146/a5bef2fc-b613-46de-8021-3e489dd663a1" width=800 />
+</p>
+<p align="center">
+  <em> Figure 2. dog6 example: the five images from the subject dog for finetuning. </em>
```
Comment (Songyuanwei) [inline examples/stable_diffusion_xl/dreambooth_finetune.md:57]: dog6 example？是多了个数字吗？

File: examples/stable_diffusion_xl/dreambooth_finetune.md [hunk 9]
```diff
@@ -59,65 +67,63 @@ Before running the fintune scripts `train_dreambooth.py`, please specify the arg
...
-Notice that the training command above gets finetuned lora weights in the specified `save_path`. Now we could use the inference command to generate images on a given prompt. Assume that the path of the trained lora weight is `output/SDXL_base_1.0_12000_lora.ckpt`, an example inference command is as
+Notice that the training command above gets finetuned lora weights in the specified `save_path`. Now we could use the inference command to generate images on a given prompt. Assume that the path of the trained lora weight is `output/SDXL_base_1.0_1000_lora.ckpt`, an example inference command is as
 
 ```shell
+export MS_PYNATIVE_GE=1
 python demo/sampling_without_streamlit.py \
   --task txt2img \
   --config configs/training/sd_xl_base_finetune_dreambooth_lora.yaml \
-  --weight checkpoints/sd_xl_base_1.0_ms.ckpt,output/SDXL_base_1.0_12000_lora.ckpt \
+  --weight checkpoints/sd_xl_base_1.0_ms.ckpt,output/SDXL_base_1.0_1000_lora.ckpt \
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/dreambooth_finetune.md:99]: is the path of lora weight below `./runs`?

File: examples/stable_diffusion_xl/dreambooth_finetune.md [hunk 10]
```diff
@@ -59,65 +67,63 @@ Before running the fintune scripts `train_dreambooth.py`, please specify the arg
...
   --instance_data_path /path/to/finetuning_data \
-  --instance_prompt "a photo of a sks dog" \
+  --instance_prompt "A photo of a sks dog" \
   --class_data_path /path/to/class_image \
-  --class_prompt "a photo of a dog" \
+  --class_prompt "A photo of a dog" \
+  --gradient_accumulation_steps 4 \
+  --ms_mode 0 \
+  --save_ckpt_interval 500 \
   --device_target Ascend
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/dreambooth_finetune.md:85]: Add distribute running commands

File: examples/stable_diffusion_xl/dreambooth_finetune.md [hunk 11]
```diff
@@ -59,65 +67,63 @@ Before running the fintune scripts `train_dreambooth.py`, please specify the arg
...
+Our implementation is trained with prior-preservation loss, which avoids overfitting and language drift. We first generate images using the pertained model with a class prompt, and input those data in parallel with our data during finetuning. The `num_class_images` in the arguments of `train_dreambooth.py`  specifies the number of class images for prior-preservation. If not enough images are present in `class_image_path`, additional images will be sampled with `class_prompt`. And you would need to relaunch the training using the command above when sampling is finished. It takes about 25 minutes to sample 50 class images on Ascend 910B.
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/dreambooth_finetune.md:88]: Ascend 910B -> Ascend

File: examples/stable_diffusion_xl/dreambooth_finetune.md [hunk 12]
```diff
@@ -59,65 +67,63 @@ Before running the fintune scripts `train_dreambooth.py`, please specify the arg
...
 
 ## Inference
 
-Notice that the training command above gets finetuned lora weights in the specified `save_path`. Now we could use the inference command to generate images on a given prompt. Assume that the path of the trained lora weight is `output/SDXL_base_1.0_12000_lora.ckpt`, an example inference command is as
+Notice that the training command above gets finetuned lora weights in the specified `save_path`. Now we could use the inference command to generate images on a given prompt. Assume that the path of the trained lora weight is `output/SDXL_base_1.0_1000_lora.ckpt`, an example inference command is as
 
 ```shell
+export MS_PYNATIVE_GE=1
 python demo/sampling_without_streamlit.py \
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/dreambooth_finetune.md:96]: default run with streamlit

File: examples/stable_diffusion_xl/gm/models/diffusion.py [hunk 13]
```diff
@@ -352,10 +352,10 @@ def do_img2img(
...
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/models/diffusion.py:355]: why delete it

**Review by zhanghuiyao — APPROVED** [review]

**Review by SamitHuang — APPROVED** [review]
